python/dgl/contrib/sampling/sampler.py

```diff
@@ -614,25 +625,233 @@ def fetch(self, current_index):
+            of head nodes with the same set of nodes uniformly randomly sampled
+            from the graph. Please see Pytorch-BigGraph for more details.
+        'PBG-tail': the negative edges are generated by corrupting a set
+            of tail nodes with the same set of nodes similar to 'PBG-head'.
+
+    The sampler returns EdgeSubgraph, where a user can access the unique head nodes
+    and tail nodes directly.
+
+    When the flag return_false_neg is turned on, the sampler will also check
+    if the generated negative edges are true negative edges and will return
+    a vector that indicates false negative edges. The vector is stored in
+    the negative graph as `false_neg` edge data.
+
+    When checking false negative edges, a user can provide edge relations
+    for a knowledge graph. A negative edge is considered as a false negative
+    edge only if the triple (source node, destination node and relation)
+    matches one of the edges in the graph.
+
+    Parameters
+    ----------
+    g : DGLGraph
+        The DGLGraph where we sample edges.
+    batch_size : int
+        The batch size (i.e, the number of edges from the graph)
+    edge_weight : tensor
+        The weight of each edge which decide the change of certain edge being sampled.
+    node_weight : tensor, optional
+        The weight of each node which decide the change of certain node being sampled.
+        Used in negative sampling. If not provided, uniform node sampling is used.
+    seed_edges : tensor, optional
+        A list of edges where we sample from.
+    shuffle : bool, optional
+        whether randomly shuffle the list of edges where we sample from.
+    num_workers : int, optional
+        The number of workers to sample edges in parallel.
+    prefetch : bool, optional
+        If true, prefetch the samples in the next batch. Default: False
+    negative_mode : string, optional
+        The method used to construct negative edges. Possible values are 'head', 'tail'.
+    neg_sample_size : int, optional
+        The number of negative edges to sample for each edge.
+    exclude_positive : int, optional
+        Whether to exclude positive edges from the negative edges.
+    return_false_neg: bool, optional
+        Whether to calculate false negative edges and return them as edge data in negative graphs.
+    relations: tensor, optional
+        relations of the edges if this is a knowledge graph.
+
+    Examples
+    --------
+    >>> for pos_g, neg_g in EdgeSampler(g, batch_size=10):
+    >>>     print(pos_g.head_nid, pos_g.tail_nid)
+    >>>     print(neg_g.head_nid, pos_g.tail_nid)
+    >>>     print(neg_g.edata['false_neg'])
+
+    Class properties
+    ----------------
+    immutable_only : bool
+        Whether the sampler only works on immutable graphs.
+        Subclasses can override this property.
+    '''
+    immutable_only = False
+
+    def __init__(
+            self,
+            g,
+            batch_size,
+            edge_weight,
```

Comment:
can we make `edge_weight` and `node_weight` be optional so we only have one EdgeSampler?

Reply:
it seems most of the logic is copied.

Reply:
Emm..
We can merge the two EdgeSample in Python side.